campaigns/workflows.md

```diff
@@ -0,0 +1,345 @@
+
+#### Possible Projects
+
+- Encoding a complete coverage of the CF conventions is too great of a task
+  for a single project, but adding functionality to
+  [CF-Xarray](https://github.com/xarray-contrib/cf-xarray) would make things
+  much easier for users.
+- Better calendar capabilities would be much appreciated.  Traditional
+  timestamps used in Pandas and Xarray are fine but they don't necessarily
+  work with non-standard calendars.  Any time a *time difference* must be
+  computed (i.e., a duration), knowledge of the calendar is critical.  The
+  Pandas and Xarray and Numpy calendar is assumed to be the proleptic
+  Gregorian calendar.  Being able to properly compute durations in other
+  calendars as well as convert between calendars would be very helpful.
+- Regridding datasets so that they can be directly compared on the same
+  grid would be very helpful.  I have always believed that a representation
+  of a gridded data object (i.e., grid + data + coordinates) should be
+  abstracted away so that the user need not know anything about the grid.
+  I call that a *field*.
+- Make it easier to plot data on unstructured grids with Python, such as MPAS
+  data.  May need utilities to deal with cyclic coordinates with unstructured
+  data.
+
+### Check-point
+
+The term *check-pointing* here refers to saving a computed artifact from
+the *Analysis* phase for later access.  Check-pointing is typically done
+specifically for long and/or expensive computations that you do not wish
+to reproduce unless the computation producing the artifact needs to
+change.  The easiest way to check-point is to simply apply an operation
+such as `to_netcdf` or `to_zarr` on an Xarray Dataset.  However, unless
+other operations in the same workflow are aware of the check-point data,
+re-running the workflow will simply recompute the artifact and re-save
+it to disk.
+
+From a user's point of view, ideally, check-pointing should be automatic.
+That is, if a an object is computed in the notebook, it is cached on a
+persistent storage platform so that it can be retrieved from disk instead
+of recomputed.  This requires knowing the *history* of the artifact data
+and knowing if any element in the artifact's history has been modified
+(e.g., a modification to a function computing an intermediate product).
+If any element in the artifact's history has changed, then it must be
+re-computed "when accessed."
+
+To limit the amount of data stored on disk, it makes sense to only
+cache selected thing in the Notebook.  These items might correspond to
+the "public API" that a user designates for a notebook in the
+*Workflow Extension* model of [Search & Discovery](#search--discovery).
+
+#### Possible Projects
+
+- An automatic caching mechanism for the data objects in the "public API"
+  of a Notebook.  Could be similar to Xpersist, but could also be more
+  generic.  This is the same as the project suggestion in
+  [Section 1](#search--discovery).
```

Comment:
For a more generic solution, I highly recommend we look into contributing to [provenance](https://github.com/bmabey/provenance). It appears to have most of the features described here, and can easily be extended to support xarray objects.

Reply:
The `provenance` package looks perfect.  I'll mention it in the text.  Thanks for the find!